### PR TITLE
check silentMode in errors on the same level as task.silent

### LINKED ISSUE #3621: silentMode in task.ts ignored for spawnNotification if response is a string
## Prerequisites

Please answer the following questions for yourself before submitting an issue. 
- [x] I am running the latest version
- [x] I checked the documentation and found no answer
- [x] I checked to make sure that this issue has not already been filed

## Check our forum:
https://forum.vuestorefront.io/

# Is there something you don't understand? What is it? Describe it.	
I am not sure if this is a bug, or if this was done intentionally. If it was intentional then I am trying to understand why.

In [core/lib/sync/task.ts#131](https://github.com/DivanteLtd/vue-storefront/blob/master/core/lib/sync/task.ts#L131) there is the following if statement. I split it up a lot to try and make it easier to read with all of the ands/ors and parentheses.

    if (
        !task.silent
        &&
        jsonResponse.result
        &&
        (
            typeof jsonResponse.result === 'string'
            ||
            (
                (
                    (
                        jsonResponse.result.result
                        ||
                        jsonResponse.result.message
                    )
                    &&
                    jsonResponse.result.code !== 'ENOTFOUND'
                )
                &&
                !silentMode
            )
        )
    )

The piece in question is the placement of `!silentMode`. Currently, if `jsonResponse.result` is a string, then `silentMode` doesn't matter and the following code is executed, which dispatches the notification to be displayed. The placement of the check on `silentMode` only matters if the response is not a string.

Why does `silentMode` not matter for string responses? Was this done for a specific reason, or was it just an oversight in the sea of parentheses that exist in this if statement?

## Additional information
This problem was noticed when testing the expiration of the Magento consumer token. If the initial responses are 401, an attempt to refresh the token is performed. Even if this is successful I still receive multiple notifications in the UI stating "Consumer is not authorized to access self" since Magento's initial response was this and it was returned as a string. But the refresh action was successful and those actions were retried successfully, so these notifications are confusing for the user.

### PR DESCRIPTION
**Pull request by gibkigonzo**
### Related issues

closes #3621

### Short description and why it's useful

with this change `silentMode` will be checked on the same level as `task.silent`

### Which environment this relates to
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [x] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [ ] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog

- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [x] I've updated the [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing VS sites with this new feature

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and currently important rules acceptance

- [x] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: core/lib/sync/task.ts [hunk 1]
```diff
@@ -128,12 +129,12 @@ function _internalExecute (resolve, reject, task: Task, currentToken, currentCar
           }
         }
 
-        if (!task.silent && jsonResponse.result && (typeof jsonResponse.result === 'string' || (((jsonResponse.result.result || jsonResponse.result.message) && jsonResponse.result.code !== 'ENOTFOUND') && !silentMode))) {
-          const message = typeof jsonResponse.result === 'string' ? jsonResponse.result : typeof jsonResponse.result.result === 'string' ? jsonResponse.result.result : jsonResponse.result.message
-
+        const errorMessage = isObject(jsonResponse.result) ? (jsonResponse.result.result || jsonResponse.result.message) : jsonResponse.result
+        const errorCode = isObject(jsonResponse.result) ? jsonResponse.result.code : jsonResponse.code
```
Comment (andrzejewsky) [inline core/lib/sync/task.ts]: I'm not to sure about this `jsonRespone.code` but let me analyse that 😄

**Comment by andrzejewsky** [comment]
try with this:

```
        const hasError = (jsonResponse) => {
          if (typeof jsonResponse.result === 'string') {
            return true
          }

          const hasMessage = jsonResponse.result.result || jsonResponse.result.message

          return hasMessage || jsonResponse.result.code !== 'ENOTFOUND'
        }

        const getErrorMessage = (jsonResponse) => {
          if (typeof jsonResponse.result === 'string') {
            return  jsonResponse.result
          }

          if (typeof jsonResponse.result.result === 'string') {
            return jsonResponse.result.result
          }

          jsonResponse.result.message
        }

        if (!task.silent && jsonResponse.result && hasError(jsonResponse) && !silentMode) {
          rootStore.dispatch('notification/spawnNotification', {
            type: 'error',
            message: i18n.t(getErrorMessage(jsonResponse)),
            action1: { label: i18n.t('OK') }
          })
        }
```

and put these function on the top of the file, now these conditions will be more readable, feel free to improve that, and please test if it works as previously. Would be nice to move this to the helpers and write some unit test

**Comment by gibkigonzo** [comment]
@andrzejewsky thanks, I've applied your suggestions :)

**Comment by rain2o** [comment]
Thanks for the fast reaction everyone! I like the cleanup of the code you did while fixing this. Looks good!

**Pull request merged**